packages/firestore/test/integration/api/snasphot_listener_source.test.ts

```diff
@@ -0,0 +1,778 @@
+          // Listen to the cache
+          const storeCacheEvent = new EventsAccumulator<QuerySnapshot>();
+          const cacheUnlisten = onSnapshot(
+            query(coll, orderBy('sort', 'asc')),
+            { source: ListenSource.Cache },
+            storeCacheEvent.storeEvent
+          );
+          let snapshot = await storeCacheEvent.awaitEvent();
+          // Cache is empty
+          expect(toDataArray(snapshot)).to.deep.equal([]);
+          expect(snapshot.metadata.fromCache).to.equal(true);
+
+          // Listen to the same query from server
+          const storeDefaultEvent = new EventsAccumulator<QuerySnapshot>();
+          const defaultUnlisten = onSnapshot(
+            query(coll, orderBy('sort', 'asc')),
+            storeDefaultEvent.storeEvent
+          );
+          snapshot = await storeDefaultEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([{ k: 'a', sort: 0 }]);
+          expect(snapshot.metadata.fromCache).to.equal(false);
+
+          // Default listener updates the cache, whish triggers cache listener to raise snapshot.
+          snapshot = await storeCacheEvent.awaitEvent();
+          expect(toDataArray(snapshot)).to.deep.equal([{ k: 'a', sort: 0 }]);
+          // The metadata is sync with server due to the default listener
+          expect(snapshot.metadata.fromCache).to.equal(false);
+
+          await storeDefaultEvent.assertNoAdditionalEvents();
+          await storeCacheEvent.assertNoAdditionalEvents();
+
+          defaultUnlisten();
+          cacheUnlisten();
+        });
+      });
+
+      it('will not get metadata only updates if listening to cache only', () => {
+        const testDocs = {
+          a: { k: 'a', sort: 0 }
+        };
+        return withTestCollection(persistence, testDocs, async coll => {
+          await getDocs(coll); // Populate the cache.
+
+          const storeEvent = new EventsAccumulator<QuerySnapshot>();
+          const unsubscribe = onSnapshot(
+            query(coll, orderBy('sort', 'asc')),
+            { includeMetadataChanges: true, source: ListenSource.Cache },
+            storeEvent.storeEvent
+          );
+
+          let snapshot = await storeEvent.awaitEvent();
+          expect(snapshot.metadata.fromCache).to.equal(true);
+          expect(toDataArray(snapshot)).to.deep.equal([{ k: 'a', sort: 0 }]);
+
+          await addDoc(coll, { k: 'b', sort: 1 });
+
+          snapshot = await storeEvent.awaitEvent();
+          expect(snapshot.metadata.hasPendingWrites).to.equal(true);
+          expect(snapshot.metadata.fromCache).to.equal(true);
+          expect(toDataArray(snapshot)).to.deep.equal([
+            { k: 'a', sort: 0 },
+            { k: 'b', sort: 1 }
+          ]);
+
+          // As we are not listening to server, the listener will not get notified
+          // when local mutation is acknowledged by server.
+          await storeEvent.assertNoAdditionalEvents();
```

Comment:
But if there is another listen to server, this one will get metadata update right? Just to clarify.

Reply:
NVM, it's the test below.